dlib_face_detection/src/face_detection_merge.cpp

```diff
@@ -0,0 +1,190 @@
+
+	std::list<PointT> point_list;
+
+	for (int i = 0; i < dFace_msg->rects.size(); ++i)
+	{
+		int x, y;
+		// center of the detected face
+		x = dFace_msg->rects[i].x + dFace_msg->rects[i].width/2;
+		y = dFace_msg->rects[i].y + dFace_msg->rects[i].height/2;
+
+		PointT point = cloud->at(x,y);
+		point_list.push_back(point);
+
+		detection.mask.roi.x = dFace_msg->rects[i].x;
+		detection.mask.roi.y = dFace_msg->rects[i].y;
+		detection.mask.roi.width = dFace_msg->rects[i].width;
+		detection.mask.roi.height = dFace_msg->rects[i].height;
+
+		detection.pose.pose.position.x = point.x;
+		detection.pose.pose.position.y = point.y;
+		detection.pose.pose.position.z = point.z;
+		detection.pose.pose.orientation.x = 0;
+		detection.pose.pose.orientation.y = 0;
+		detection.pose.pose.orientation.z = 0;
+		detection.pose.pose.orientation.w = 1;
+
+		detArray->detections.push_back(detection);
+	}
+
+
+	for (int i = 0; i < hbFace_msg->rects.size(); ++i)
+	{
+		int x, y;
+		// center of the detected face
+		x = hbFace_msg->rects[i].x + hbFace_msg->rects[i].width/2;
+		y = hbFace_msg->rects[i].y + hbFace_msg->rects[i].height/2;
+
+		PointT point = cloud->at(x,y);
+
+		bool det_flag = false;
+		// check if the face was detected previously by direct face detection
+		if (point_list.size() == 0)
+			det_flag = true;
+		else
+		{
+			std::list<PointT>::iterator it = point_list.begin();
+			for (int i = 0; i < point_list.size(); ++i)
+			{
+				if (abs(point.x - it->x) > 0.1 && abs(point.z - it->z) > 0.1)
+					det_flag = true;
+				else
+				{
+					det_flag = false;
+					break;
+				}
+				++it;
+			}
+		}
+
+		if(det_flag)
+		{
+
+			PointT point = cloud->at(x,y);
+
+			detection.mask.roi.x = hbFace_msg->rects[i].x;
+			detection.mask.roi.y = hbFace_msg->rects[i].y;
+			detection.mask.roi.width = hbFace_msg->rects[i].width;
+			detection.mask.roi.height = hbFace_msg->rects[i].height;
+
+			detection.pose.pose.position.x = point.x;
+			detection.pose.pose.position.y = point.y;
+			detection.pose.pose.position.z = point.z;
+			detection.pose.pose.orientation.x = 0;
+			detection.pose.pose.orientation.y = 0;
+			detection.pose.pose.orientation.z = 0;
+			detection.pose.pose.orientation.w = 1;
+
+			detArray->detections.push_back(detection);
+		}
+
+	}
+
+	face_pub_.publish(detArray);
+
+	if (enable_tf_ )
+		tfpublisher(detArray);
+
+}
+
+void FaceDetectionMerge::tfpublisher(const cob_perception_msgs::DetectionArray::ConstPtr detArray)
+{
+	geometry_msgs::PoseStamped pose_world;
+
+	int k = 0;
+
+	for (int i = 0; i < detArray->detections.size(); i++)
+	{
+		std::stringstream s;
+		k += 1;
+		s << "dlib_face_" << k;
+
+		int a = 0;
+		while (a < 5)
+		{
+			// transform pose to "/odom_combined" (world coordinates)
+			try
+			{
+				ros::Time now = ros::Time(0);
+				tfl_.waitForTransform("/odom_combined", detArray->header.frame_id, now, ros::Duration(0.5));
+				tfl_.transformPose("/odom_combined", now, detArray->detections[i].pose, "/odom_combined", pose_world);
```

Comment:
do we really want to transform against `odom_combined`? maybe `base_link` is better as it is "on" the robot and not affected by navigation/localization...